deeptime/src/include/deeptime/markov/msm/tram/tram.h

```diff
@@ -90,6 +101,122 @@ class ExchangeableArray {
 };
 
 
+// natural logarithm of the statistical weight per sample, log \mu^k(x).
+// If thermState =-1, this is the unbiased statistical sample weight, log \mu(x).
+template<typename dtype>
+static const std::vector<dtype> computeSampleWeightsLog(StateIndex thermStateIndex,
+                                    const DTraj & dtraj,
+                                    const BiasMatrix<dtype> & biasMatrix,
+                                    const np_array<dtype> & thermStateEnergies,
+                                    const np_array_nfc<dtype> & modifiedStateCountsLog) {
+    std::size_t nThermStates = static_cast<StateIndex>(thermStateEnergies.size());
+
+    std::vector<dtype> sampleWeights(dtraj.size());
+    std::vector<dtype> scratch(nThermStates);
+
+    ArrayBuffer<DTraj, 1> dtrajBuf{dtraj};
+    ArrayBuffer<BiasMatrix<dtype>, 2> biasMatrixBuf{biasMatrix};
+    ArrayBuffer<np_array_nfc<dtype>, 2> modifiedStateCountsLogBuf{modifiedStateCountsLog};
+    ArrayBuffer<np_array_nfc<dtype>, 1> thermStateEnergiesBuf{thermStateEnergies};
+
+    for (auto x = 0; x < dtraj.size(); ++x) {
+        auto i = dtrajBuf(x);
+        if (i < 0) {
+            // The sample weight has no meaning, as it does not exist in the connected set. The value is set to -inf
+            // so that the weight is zero and the sample does not contribute to the output when computing observables.
+            sampleWeights[x] = -std::numeric_limits<dtype>::infinity();
+            continue;
+        }
+        int o = 0;
+        for (StateIndex l = 0; l < nThermStates; ++l) {
+            if (modifiedStateCountsLogBuf(l, i) > -std::numeric_limits<dtype>::infinity()) {
+                scratch[o++] = modifiedStateCountsLogBuf(l, i) - biasMatrixBuf(x, l);
+            }
+        }
+        auto log_divisor = numeric::kahan::logsumexp_sort_kahan_inplace(scratch.begin(), o);
+        if (thermStateIndex == -1) {// get unbiased sample weight
+            sampleWeights[x] = -log_divisor;
+        } else { // get biased sample weight for given thermState index
+            sampleWeights[x] = -biasMatrixBuf(x, thermStateIndex) - log_divisor
+                                        + thermStateEnergiesBuf(thermStateIndex);
+        }
```

Comment:
instead of if -> continue I think if - else is a bit more readable (`goto bad;`!)